scripts/multiplayer.gd

```diff
@@ -0,0 +1,179 @@
+#TO DO: Multiplayer with more than 2 players
+
+extends Panel
+
+const SERVER_PORT = 25700
+const MAX_PLAYERS = 2
+var SERVER_IP = "127.0.0.1"
+var player_info = {}
+var username
+onready var username_text = $username
+onready var ip_adress = $ip_adress
+onready var waiting_text = $waiting/waiting_text
+onready var waiting = $waiting
+var mp_type
+onready var game = $game
+const romp_online = preload("res://levels/romp_online.tscn")
+const dummy_scene = preload("res://actors/dummy.tscn")
+const player_scene = preload("res://actors/player.tscn")
+var game_running = false
+
+func _on_Host_pressed():
+	start_multiplayer("server")
+
+func _on_Join_pressed():
+	if username_text.text == "":
+		#Abort if no username has been entered.
+		username_text.get_node("warning").visible = true
+		return
+	ip_adress.visible = true
+	ip_adress.get_node("ip_adress_animation").play("ip")
+
+func _on_BackIP_pressed():
+	ip_adress.visible = false
+
+func _on_Connect_pressed():
+	if ip_adress.get_node("LineEdit").text == "":
+		return
+	ip_adress.visible = false
+	SERVER_IP = ip_adress.get_node("LineEdit").text
+	start_multiplayer("client")
+
+func _on_BackButton_pressed():
+	get_tree().network_peer = null
+	get_tree().change_scene("res://menus/playmenu.tscn")
+
+func _process(delta):
+	if game_running:
+		#Update player position
+		var character = game.get_node("romp_online/player/Player/KinematicBody")
+		rpc_unreliable("position_update", character.global_transform.origin, character.rotation)
+		if mp_type != "server":
+			return
+		#The server tells the clients the ball positions
+		var targets = game.get_node("romp_online/Targets")
+		var balls_pos_array = []
+		for i in targets.get_children():
+			balls_pos_array.append(i.global_transform.origin)
+		rpc_unreliable("targets_update", balls_pos_array)
+
+remote func targets_update(pos_array):
+	var targets = game.get_node("romp_online/Targets")
+	for i in targets.get_child_count():
+		targets.get_children()[i].global_transform.origin = pos_array[i]
+
+remote func position_update(pos, rot):
+	if game.get_node("romp_online/dummies") == null:
+		return
+	var actor = game.get_node("romp_online/dummies").get_node(str(get_tree().get_rpc_sender_id()))
+	if actor != null:
+		actor.translation = pos
+		actor.rotation = rot
+
+func start_multiplayer(type):
+	if username_text.text == "":
+		#Abort if no username has been entered.
+		username_text.get_node("warning").visible = true
+		return
+	username = username_text.text
+	
+	#Show waiting screen
+	waiting.visible = true
+	mp_type = type
+	
+	#Create new ENet
+	var peer = NetworkedMultiplayerENet.new()
+	match type:
+		"server":
+			peer.create_server(SERVER_PORT, MAX_PLAYERS)
+			waiting_text.text = "Waiting for players..."
+		"client":
+			peer.create_client(SERVER_IP, SERVER_PORT)
+			waiting_text.text = "Connecting..."
+	get_tree().network_peer = peer
```

Comment:
My only issue with this is that there's no way to get out of connecting to a server if you typed the IP in wrong, or your host wasn't ready. If you're a host, and you get no clients, you also have no way to get out of waiting. This leaves you forced to <kbd>ALT + F4</kbd>.

Granted, I'm glad this works at all, I think this is an amazing feature, I just hesitate if you can get into a game like this but you can't get out..

As a client, typing in random IPs gives me errors for create_client. Some way to check for this error (does `create_client()` return an errorcode?) would be nice to prevent this and boot the player back to the previous screen.

```
E 0:09:12.245   _resolve_hostname: getaddrinfo failed! Cannot resolve hostname.
  <C++ Source>  drivers/unix/ip_unix.cpp:112 @ _resolve_hostname()
  <Stack Trace> multiplayer.gd:91 @ start_multiplayer()
                multiplayer.gd:40 @ _on_Connect_pressed()
```